src/transformation/context/context.ts

```diff
@@ -49,80 +53,92 @@ export class TransformationContext {
             this.diagnostics.push(unsupportedNodeKind(node, node.kind));
             return isExpression ? [lua.createNilLiteral()] : [];
         }
 
         const previousNodeVisitors = this.currentNodeVisitors;
-        this.currentNodeVisitors = [...nodeVisitors];
+        const previousNodeVisitorsIndex = this.currentNodeVisitorsIndex;
+        this.currentNodeVisitors = nodeVisitors;
+        this.currentNodeVisitorsIndex = nodeVisitors.length - 1;
 
-        const visitor = this.currentNodeVisitors.pop()!;
-        const result = unwrapVisitorResult(visitor.transform(node, this));
+        const visitor = this.currentNodeVisitors[this.currentNodeVisitorsIndex];
+        const result = visitor(node, this);
 
         this.currentNodeVisitors = previousNodeVisitors;
+        this.currentNodeVisitorsIndex = previousNodeVisitorsIndex;
 
         return result;
     }
 
     public superTransformNode(node: ts.Node): lua.Node[] {
-        if (this.currentNodeVisitors.length === 0) {
+        return unwrapVisitorResult(this.doSuperTransformNode(node));
+    }
+
+    private doSuperTransformNode(node: ts.Node): OneToManyVisitorResult<lua.Node> {
+        if (--this.currentNodeVisitorsIndex < 0) {
             throw new Error(`There is no super transform for ${ts.SyntaxKind[node.kind]} visitor`);
         }
 
-        const visitor = this.currentNodeVisitors.pop()!;
-        return unwrapVisitorResult(visitor.transform(node, this));
+        const visitor = this.currentNodeVisitors[this.currentNodeVisitorsIndex];
+        return unwrapVisitorResult(visitor(node, this));
     }
 
     public transformExpression(node: ExpressionLikeNode): lua.Expression {
-        const [result] = this.transformNode(node, true);
+        const result = this.transformNodeRaw(node, true);
+        return this.assertIsExpression(node, result);
+    }
 
+    private assertIsExpression(node: ExpressionLikeNode, result: OneToManyVisitorResult<lua.Node>): lua.Expression {
         if (result === undefined) {
             throw new Error(`Expression visitor for node type ${ts.SyntaxKind[node.kind]} did not return any result.`);
         }
-
+        if (Array.isArray(result)) {
+            return result[0] as lua.Expression;
+        }
         return result as lua.Expression;
     }
 
     public superTransformExpression(node: ExpressionLikeNode): lua.Expression {
-        const [result] = this.superTransformNode(node);
-
-        if (result === undefined) {
-            throw new Error(`Expression visitor for node type ${ts.SyntaxKind[node.kind]} did not return any result.`);
-        }
-
-        return result as lua.Expression;
+        const result = this.doSuperTransformNode(node);
+        return this.assertIsExpression(node, result);
     }
 
     public transformStatements(node: StatementLikeNode | readonly StatementLikeNode[]): lua.Statement[] {
-        return castArray(node).flatMap(n => {
+        return castArrayFlatMap(node, n => {
```

Comment:
I just did a remeasurment; as this method is commonly called, this change provide a consistently measurable (30ms over a 5500ms time) speedup. If the slight clarity is still prefered over the slight improvement, this could be reverted.

Reply:
I don't think this extra code is worth for 0.4% performance increase, please revert

Reply:
Reverted.